Implement inertial forces for shell element

recuero opened the pull request:
This PR implements and initially verifies the inertial forces of MOOSE's shell element. 

The inertia class inherits from `ADTimeKernel` and overrides `ComputeResidual` and `computeADOffDiagJacobian` to avoid double looping over the element quadrature points. 

Initial verification is done on a single element example checking the fundamental natural frequency.

##
- [x] Initial implementation
- [x] Verification on simple problem: Fundamental frequency
- [x] Newmark-beta, Rayleigh damping inertial forces
- [x] HHT inertial forces
- [x] Tests ensuring objectivity (partial)

Refs. #15067

recuero committed:
Initial commit. Doesn't compile or do anything.

recuero committed:
Preliminary (wrong) residual for rotational degrees of freedom.

recuero committed:
Some intermediate operations in compilable form.

recuero committed:
Make it AD. Compiles.

recuero committed:
One more commit to switch branches.

recuero committed:
Use inefficient helper function to construct the Jacobian.

recuero committed:
Temp change

recuero committed:
Next step: Figure out computeJacobian()

recuero committed:
Shell vibrates. Need to verify results are okay and remove a myriad of debug comments

recuero committed:
Verified for bending vibrations

recuero committed:
Add two tests, clean up, add some documentation. Refs. #15067.

recuero committed:
Label tests as heavy Refs. #15067

recuero committed:
Fix bad character in TeX bib file

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/501158/) on 22e9791 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15068/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15068/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 832ea0a34942120ac5688110804f533c20ff2f3e`

recuero committed:
Apply clang format

recuero committed:
Change file names

moosebuild commented:
Job [App tests (Non Conda)](https://civet.inl.gov/job/501390/) on 28727ce : invalidated by @recuero

recuero committed:
Fix rotation matrix modifications. It acts as a deformation gradient (also stretches/deforms)

Created a new material property with a rotation matrix that only rotates for use in the Inertia kernel.

recuero committed:
Add missing mesh file

recuero committed:
Update Bib again

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/502031/) on 6d25ebb wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15068/site)

This comment will be updated on new commits.

recuero committed:
Update shell element inertia code to AD 2.0

recuero committed:
Fix BCs in inclined shell test

recuero committed:
Update reference test file for inclined shell -- dynamics

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/505948/) on c241ed1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15068/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15068/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c3a1cdfe44c215caa66fc4c6243038659a748cf5`

recuero committed:
Apply clang format

recuero committed:
Change input files to avoid AD/non-AD clashes

recuero committed:
Enable HHT/Rayleigh damping for shell inertial forces.

recuero committed:
Clean up

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/507481/) on f4ef4b1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15068/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15068/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 77b75378dd3cd3e4eff8741383a1ff9f847c3ded`

recuero committed:
Provide 'detail' entry for all tests

moosebuild commented:
Job [Test debug PETSc alt](https://civet.inl.gov/job/507591/) on ea3eda1 : invalidated by @recuero

recuero commented:
@dschwen please take a look

dschwen commented on modules/tensor_mechanics/include/kernels/ADInertialForceShell.h:
```diff
@@ -0,0 +1,186 @@
+#include "RankTwoTensor.h"
+#include "libmesh/dense_vector.h"
+
+// Forward Declarations
+class ADInertialForceShell;
```
These line are not needed. Please remove

dschwen commented on modules/tensor_mechanics/include/kernels/ADInertialForceShell.h:
```diff
@@ -0,0 +1,186 @@
+  /**
+   * Rotational transformation from global to initial beam local
+   * coordinate system
+   **/
```
Just one star in closing the doxygen comment

dschwen commented on modules/tensor_mechanics/include/kernels/ADInertialForceShell.h:
```diff
@@ -0,0 +1,186 @@
+   **/
+  ADRealVectorValue _vel_0, _vel_1, _rot_vel_0, _rot_vel_1;
+  ADRealVectorValue _vel_2, _vel_3, _rot_vel_2, _rot_vel_3;
+  /**
```
empty line above

dschwen commented on modules/tensor_mechanics/include/kernels/ADInertialForceShell.h:
```diff
@@ -0,0 +1,186 @@
+  /// Direction along which residual is calculated
+  const unsigned int _component;
+
+  /**
+   * Current translational and rotational velocities at the two nodes
+   * of the beam in the global coordinate system
+   **/
+  ADRealVectorValue _vel_0, _vel_1, _rot_vel_0, _rot_vel_1;
+  ADRealVectorValue _vel_2, _vel_3, _rot_vel_2, _rot_vel_3;
```
The comment will only be applied to the first declatation of member variables. I suggest to put all in one line, or all in separate lines and then use the

```
/// @{ Comment that applies to all members below, until the closing comment
Type1 member1;
Type2 member2;
...
///@}
```

dschwen commented on modules/tensor_mechanics/include/kernels/ADInertialForceShell.h:
```diff
@@ -0,0 +1,186 @@
+  ADRealVectorValue _local_vel_0, _local_vel_1, _local_rot_vel_0, _local_rot_vel_1;
+  ADRealVectorValue _local_vel_2, _local_vel_3, _local_rot_vel_2, _local_rot_vel_3;
+  /**
+   * Current translational and rotational velocities at the two nodes
```
Current? No, old.

dschwen commented on modules/tensor_mechanics/include/kernels/ADInertialForceShell.h:
```diff
@@ -0,0 +1,186 @@
+   * of the beam in the initial beam local coordinate system
+   **/
+  ADRealVectorValue _local_accel_0, _local_accel_1, _local_rot_accel_0, _local_rot_accel_1;
+  ADRealVectorValue _local_accel_2, _local_accel_3, _local_rot_accel_2, _local_rot_accel_3;
```
You have a ton of those sets. I would strongly suggest to make a struct that holds your 3 position and rotation vector quantities!

```
struct PosRotVectors
{
   std::array<ADRealVectorValue, 3> pos;
   std::array<ADRealVectorValue, 3> rot;
};
```

then you'd have 

```
...
/// Current velocities at the two nodes in the initial beam local coordinate system
PosRotVectors _local_vel;

/// Old velocities at the two nodes in the initial beam local coordinate system
PosRotVectors _local_old_vel;

/// Current accelerations at the two nodes in the initial beam local coordinate system
PosRotVectors _local_accel;
```

easier to read. Then `_local_rot_accel_1 -> _local_accel.rot[1]`

dschwen commented on modules/tensor_mechanics/include/kernels/ADInertialForceShell.h:
```diff
@@ -0,0 +1,186 @@
+  ADRealVectorValue _global_force_0, _global_force_1, _global_moment_0, _global_moment_1;
+  ADRealVectorValue _global_force_2, _global_force_3, _global_moment_2, _global_moment_3;
+
+  // AMR
```
remove

dschwen commented on modules/tensor_mechanics/include/kernels/ADInertialForceShell.h:
```diff
@@ -0,0 +1,186 @@
+  /// Normal to the element at the 4 nodes.
+  std::vector<ADRealVectorValue> _node_normal;
+
+  /// Node 1 g vector in reference configuration (mass matrix)
```
this should be a single `std::array<ADRealVectorValue, 4> _0g_vector;` (not the one-off indexing)

recuero replied:
There are two vectors at each node.  I kept them separate but made them an array:

`std::array<ADRealVectorValue, 2> _0g1_vectors` 

The one-off indexing is just meant to preserve the notation from papers/PhD thesis. Rest of the comments should be fully addressed.

recuero committed:
Address reviewer's comments

dschwen commented on modules/tensor_mechanics/include/kernels/ADInertialForceShell.h:
```diff
@@ -0,0 +1,172 @@
+   * coordinate system
+   */
+  ADRealVectorValue _global_force_0, _global_force_1, _global_moment_0, _global_moment_1;
+  ADRealVectorValue _global_force_2, _global_force_3, _global_moment_2, _global_moment_3;
```
those could be std::arrays as well...

dschwen commented on modules/tensor_mechanics/src/materials/ADComputeIncrementalShellStrain.C:
```diff
@@ -341,6 +344,24 @@ ADComputeIncrementalShellStrain::computeGMatrix()
 
+      double normx = std::sqrt(J(0, 0) * J(0, 0) + J(0, 1) * J(0, 1) + J(0, 2) * J(0, 2));
+      double normy = std::sqrt(J(1, 0) * J(1, 0) + J(1, 1) * J(1, 1) + J(1, 2) * J(1, 2));
+      double normz = std::sqrt(J(2, 0) * J(2, 0) + J(2, 1) * J(2, 1) + J(2, 2) * J(2, 2));
```
Please use `Real` rather than `double`

dschwen commented on modules/tensor_mechanics/test/tests/shell/dynamics/shell_dynamics_bending_moment_free_orientation_inclined_hht.i:
```diff
@@ -0,0 +1,481 @@
+
+  # aux variables for dynamics
+  [./vel_x]
+  order = FIRST
+  family = LAGRANGE
```
Those are the default values. I'd just remove those.

recuero committed:
Address reviewer's comments II

recuero commented:
@dschwen please take another look

dschwen reviewed (approved):
looks good now
